### ISSUE TITLE
How to deal with mixed types from WordPress core?

### ISSUE THREAD

**Issue by claytoncollie** [szepeviktor/phpstan-wordpress#29]
When I use functions from WordPress, PHPstan is flagging them as having incorrect mixed types when used in `printf` or `sprintf` functions.

First example is: `esc_url( get_permalink() )` which says `Parameter #1 $url of function esc_url expects string, string|false given.` because `get_permalink` return string or integer according to the docs.

Second example is: `get_the_term_list( get_the_ID(), 'category', '', ', ', '' )` which says `Parameter #1 $post_id of function get_the_term_list expects int, int|false given.` because `get_the_ID()` returns an integer or false according to the docs.

Am I doing something wrong when using these functions or do I need to mark these lines as OK to ignore?

Here is my entire block of code.

```php
$output .= sprintf(
	'<li><h3 class="entry-title"><a href="%s">%s</a></h3>%s<span class="entry-sep">&middot;</span>%s<span class="entry-sep">&middot;</span>%s</li>',
	esc_url( get_permalink() ),
	esc_html( get_the_title() ),
	get_the_date(),
	get_the_term_list( get_the_ID(), 'category', '', ', ', '' ),
	poa_get_the_time_to_read(),
);
```

**Comment by szepeviktor** [comment]
> `esc_url( get_permalink() )` which says `Parameter #1 $url of function esc_url expects string, string|false given.` because `get_permalink` return string or integer according to the docs.

Yes, it is right. False will not fit into printf.
Please consider checking the output.

**Comment by szepeviktor** [comment]
...or for the type

```php
/** @var string $myDearForcedTypedVar */
$myDearForcedTypedVar = get_permalink();
```

But you haven't heard it from me!

**Comment by szepeviktor** [comment]
Consider Timber v2.

You heard it from me :)

**Comment by claytoncollie** [comment]
I understand now. `printf` and `sprintf` have their own outputs and the parameter is colliding with that. We rely heavily on those functions so I will ignore these warnings from phpstan. Thanks for the help.

**Comment by szepeviktor** [comment]
@claytoncollie Can you think of a situation when `get_permalink` returns `false`?

**Comment by claytoncollie** [comment]
I cannot think of the situation for `get_permalink` to be `false` but we cant control the function or docs since it is in WP core. I think this is more of a `printf` and `sprintf` outputs colliding with `get_permalink` outputs. Neither is wrong but they don't play nicely.

**Comment by szepeviktor** [comment]
Docs say

> The permalink URL or false if post does not exist.

I leave it up to you. Robust code is expensive and robustness could only be detected in couple of months of monitoring.

**Comment by claytoncollie** [comment]
I agree with you that robustness should be the goal. I don't know how to get there without the work around you presented above in many different places.

1. Make my own function with its own docblock
```
/** @var string $myDearForcedTypedVar */
$myDearForcedTypedVar = get_permalink();
```

2. Use timber. (not going down that path)

3. Insert `global $post;` at the top of my functions.

Does that sound like the possible solutions?

**Comment by szepeviktor** [comment]
WordPress has some substantial problems.
The main one being it lacks templating/routing.
PHPStan is for OOP code. False-returning functions are not OOP. A problem should be [reported in an exception](https://github.com/thecodingmachine/safe) in OOP.

I am not able to resolve them. I am sorry.

**Comment by szepeviktor** [comment]
But! You can add custom stubs to PHPStan https://phpstan.org/user-guide/stub-files

```php
/** @return string */
function get_permalink() {}
```

**Comment by claytoncollie** [comment]
I think I finally understand this concept now. If a function returning mixed types, I have to check the type and return what the next function needs. In this case, `esc_url` only accepts a `string` value so we need to make sure we send it a string in any case, true or false or empty or null.

```
$permalink = get_permalink();
$permalink = $permalink ? $permalink : '';
echo esc_url( $permalink);
```

Take a little extra code but this is more type safe than `echo esc_url( get_permalink() );`

**Comment by szepeviktor** [comment]
Yes.

1. get_permalink
2. check its return value
3. give it to esc_url
